### ISSUE TITLE
About FP16 nan problem?

### ISSUE THREAD

**Issue by ElliottYan** [OpenNMT/OpenNMT-py#1796]
Hi, 

I'm trying to reproduce some results for Transformer-base with WMT'14 EN-DE dataset. However, when I trained a transformer with relative position information, this error occurs. It has successfully trained for about 5000 steps, but suddenly it collapsed with nan.

![image](https://user-images.githubusercontent.com/10862038/82431330-1763bf00-9ac1-11ea-84fa-4133b3746f7f.png)

The apex is installed in the way described by https://github.com/OpenNMT/OpenNMT-py/issues/1645#issuecomment-555405410 and I also used fusedadam.

Can somebody help me over here? Thanks in advance.

**Comment by vince62s** [comment]
Please post your command line.
Also let us know the Pytorch version + Apex git tag.

**Comment by AbelCui** [comment]
Hi, 

I also met same problem when I use Transformer with adam, without relative position information, and my Pytorch is 1.4 with CUDA 10.0 and Apex git branch is master, commit: 8abb6908e3b4f0c1c2cd516c45557e935d6357eb.

Is there somebody can solve this problem? Thanks.

**Comment by AbelCui** [comment]
And I run onmt (1.1.1) training on 4 V100 GPUs with these options:
save_checkpoint_steps: 5000
keep_checkpoint: 20
seed: 3435
train_steps: 500000
valid_steps: 5000
warmup_steps: 8000
report_every: 100

model_dtype: fp16
decoder_type: transformer
encoder_type: transformer
enc_layers: 12
dec_layers: 6
rnn_size: 1024
word_vec_size: 1024
transformer_ff: 4096
heads: 16

accum_count: 4
optim: adam
adam_beta1: 0.9
adam_beta2: 0.98
decay_method: noam
learning_rate: 2
learning_rate_decay: 1.0
max_grad_norm: 0.0

batch_size: 8192
batch_type: tokens
normalization: tokens
dropout: 0.1
attention_dropout: 0.1
label_smoothing: 0.1

max_generator_batches: 2

param_init: 0.0
param_init_glorot: 'true'
position_encoding: 'true'

world_size: 4
gpu_ranks:
- 0
- 1
- 2
- 3

**Comment by vince62s** [comment]
in FP16 you need max_generator_batches: 2
let me know if it helps.

**Comment by AbelCui** [comment]
I rebuild a new docker and it works.

**Comment by gaussmao** [comment]
I have a similar issue as mentioned above, and I've submitted it on the OpenNMT forum (https://forum.opennmt.net/t/about-the-issues-with-opennmt-py-in-machine-translation-models-really-looking-forward-to-the-experts-response-and-assistance-many-thanks/5634). Do you have time to help clarify it for me? Thank you very much!
